File: sclib.js/smd/SMDTransformer.js
import { ALPHABET } from './../constants.js';

export class SMDTransformer {
    static transform_array(array, transformFunction) {
        const stack = [[array, []]];
        const transformed = {};
   
        while (stack.length > 0) {
            const [current, path] = stack.pop();
   
            for (const [key, value] of Object.entries(current)) {
                const newKey = transformFunction(key);
                const currentPath = [...path, newKey];
                
                if (value && typeof value === 'object' && !Array.isArray(value)) {
                    stack.push([value, currentPath]);
                    let ref = transformed;
                    for (const pathKey of currentPath) {
                        if (!(pathKey in ref)) {
                            ref[pathKey] = {};
                        }
                        ref = ref[pathKey];
                    }
                } else {
                    let ref = transformed;
                    const lastKey = currentPath.pop();
                    for (const pathKey of currentPath) {
                        if (!(pathKey in ref)) {
                            ref[pathKey] = {};
                        }
                        ref = ref[pathKey];
                    }
                    ref[lastKey] = value;
                }
            }
        }
   
        return transformed;
    }

    static sort_object(array) {
        const stack = [[array, []]];
        const sorted = {};
   
        while (stack.length > 0) {
            const [current, path] = stack.pop();
            const sortedKeys = Object.keys(current).sort();
   
            for (const key of sortedKeys) {
                const value = current[key];
                const currentPath = [...path, key];
                
                if (value && typeof value === 'object' && !Array.isArray(value)) {
                    let ref = sorted;
                    stack.push([value, currentPath]);

                    for (const pathKey of currentPath) {
                        if (!(pathKey in ref)) {
                            ref[pathKey] = {};
                        }

                        ref = ref[pathKey];
                    }
                } else {
                    let ref = sorted;
                    const lastKey = currentPath.pop();

                    for (const pathKey of currentPath) {
                        if (!(pathKey in ref)) {
                            ref[pathKey] = {};
                        }

                        ref = ref[pathKey];
                    }

                    ref[lastKey] = value;
                }
            }
        }
   
        return sorted;
    }
   
    static shorten(data) {
        if (typeof data === 'string') {
            const index = Object.entries(ALPHABET).find(([_, value]) => value === data)?.[0];
            return index !== undefined ? index : data;
        }
   
        const transformed = this.transform_array(data, (key) => {
            const index = Object.entries(ALPHABET).find(([_, value]) => value === key)?.[0];
            return index !== undefined ? index : key;
        });
   
        return this.sort_object(transformed);
    }
   
    static expand(data) {
        if (typeof data === 'string') {
            return ALPHABET[data] || data;
        }
   
        const transformed = this.transform_array(data, (key) => {
            return ALPHABET[key] || key;
        });
   
        return this.sort_object(transformed);
    }
}

File: pysclib/smd_transformer.py
from constants import ALPHABET

class smd_transformer:
    @staticmethod
    def transform_array(array, transform_function):
        stack = [(array, [])]
        transformed = {}

        while stack:
            current, path = stack.pop()

            for key, value in current.items():
                new_key = transform_function(key)
                current_path = path + [new_key]

                if isinstance(value, dict):
                    stack.append((value, current_path))
                    ref = transformed
                    for path_key in current_path:
                        if path_key not in ref:
                            ref[path_key] = {}
                        ref = ref[path_key]
                else:
                    ref = transformed
                    last_key = current_path.pop()
                    for path_key in current_path:
                        if path_key not in ref:
                            ref[path_key] = {}
                        ref = ref[path_key]
                    ref[last_key] = value

        return transformed

    @staticmethod
    def sort_object(array):
        stack = [(array, [])]
        sorted_obj = {}

        while stack:
            current, path = stack.pop()
            sorted_keys = sorted(current.keys())

            for key in sorted_keys:
                value = current[key]
                current_path = path + [key]

                if isinstance(value, dict):
                    ref = sorted_obj
                    stack.append((value, current_path))

                    for path_key in current_path:
                        if path_key not in ref:
                            ref[path_key] = {}
                        ref = ref[path_key]
                else:
                    ref = sorted_obj
                    last_key = current_path.pop()

                    for path_key in current_path:
                        if path_key not in ref:
                            ref[path_key] = {}
                        ref = ref[path_key]

                    ref[last_key] = value

        return sorted_obj

    @staticmethod
    def shorten(data):
        if isinstance(data, str):
            index = next((k for k, v in ALPHABET.items() if v == data), None)
            return index if index is not None else data

        transformed = smd_transformer.transform_array(data, lambda key: next((k for k, v in ALPHABET.items() if v == key), key))
        return smd_transformer.sort_object(transformed)

    @staticmethod
    def expand(data):
        if isinstance(data, str):
            return ALPHABET.get(data, data)

        transformed = smd_transformer.transform_array(data, lambda key: ALPHABET.get(key, key))
        return smd_transformer.sort_object(transformed)